metadata-ingestion/src/datahub/ingestion/source/sql/snowflake.py

```diff
@@ -82,53 +83,193 @@ def __init__(self, config: SnowflakeConfig, ctx: PipelineContext):
         self.provision_role_in_progress: bool = False
         self.profile_candidates: Dict[str, List[str]] = {}
 
+    @staticmethod
+    def check_capabilities(  # noqa: C901
+        conn: connector.SnowflakeConnection, connection_conf: SnowflakeConfig
+    ) -> Dict[Union[SourceCapability, str], CapabilityReport]:  # noqa: C901
+
+        # Currently only overall capabilities are reported.
+        # Resource level variations in capabilities are not considered.
+
+        @dataclass
+        class SnowflakePrivilege:
+            privilege: str
+            object_name: str
+            object_type: str
+
+        def query(query):
+            logger.info("Query : {}".format(query))
+            resp = conn.cursor().execute(query)
+            return resp
+
+        _report: Dict[Union[SourceCapability, str], CapabilityReport] = dict()
+        privileges: List[SnowflakePrivilege] = []
+        capabilities: List[SourceCapability] = [c.capability for c in SnowflakeSource.get_capabilities() if c.capability not in (SourceCapability.PLATFORM_INSTANCE, SourceCapability.DOMAINS, SourceCapability.DELETION_DETECTION)]  # type: ignore
+
+        cur = query("select current_role()")
+        current_role = [row[0] for row in cur][0]
+
+        cur = query("select current_secondary_roles()")
+        secondary_roles_str = json.loads([row[0] for row in cur][0])["roles"]
+        secondary_roles = (
+            [] if secondary_roles_str == "" else secondary_roles_str.split(",")
+        )
+
+        roles = [current_role] + secondary_roles
+
+        # PUBLIC role is automatically granted to every role
+        if "PUBLIC" not in roles:
+            roles.append("PUBLIC")
+        i = 0
+
+        while i < len(roles):
+            role = roles[i]
+            i = i + 1
+            try:
+                cur = query(f"show grants to role {role}")
+            except connector.errors.ProgrammingError:
+                # for some roles, quoting is necessary. for example test-role
+                cur = query(f'show grants to role "{role}"')
```

Comment:
why not use this query by default?

Reply:
yes, it can be quoted by default. I will make this change.